rosidl_generator_c/src/u16string_functions.c

```diff
@@ -0,0 +1,197 @@
+  } else {
+    /* ensure that data, size, and capacity values are consistent */
+    if (0 != str->size) {
+      fprintf(stderr, "Unexpected condition: string size was non-zero for deallocated data! "
+        "Exiting.\n");
+      exit(-1);
+    }
+    if (0 != str->capacity) {
+      fprintf(stderr, "Unexpected behavior: string capacity was non-zero for deallocated data! "
+        "Exiting.\n");
+      exit(-1);
+    }
+  }
+}
+
+bool
+rosidl_generator_c__U16String__assignn(
+  rosidl_generator_c__U16String * str, const uint16_t * value, size_t n)
+{
+  if (!str) {
+    return false;
+  }
+  // a NULL value is not valid
+  if (!value) {
+    return false;
+  }
+  // since n + 1 bytes are being allocated n can't be the maximum value
+  if (n == SIZE_MAX) {
+    return false;
+  }
+  uint16_t * data = realloc(str->data, (n + 1) * sizeof(uint16_t));
+  if (!data) {
+    return false;
+  }
+  memcpy(data, value, n * sizeof(uint16_t));
+  data[n] = 0;
+  str->data = data;
+  str->size = n;
+  str->capacity = n + 1;
+  return true;
+}
+
+bool
+rosidl_generator_c__U16String__assign(
+  rosidl_generator_c__U16String * str, const uint16_t * value)
+{
+  return rosidl_generator_c__U16String__assignn(
+    str, value, rosidl_generator_c__U16String__len(value));
+}
+
+size_t
+rosidl_generator_c__U16String__len(const uint16_t * value)
+{
+  if (!value) {
+    return 0;
+  }
+  size_t len = 0;
+  while (value[len++]) {}
+  return len;
+}
+
+bool
+rosidl_generator_c__U16String__Sequence__init(
+  rosidl_generator_c__U16String__Sequence * sequence, size_t size)
+{
+  if (!sequence) {
+    return false;
+  }
+  rosidl_generator_c__U16String * data = NULL;
+  if (size) {
+    data = (rosidl_generator_c__U16String *)calloc(size, sizeof(rosidl_generator_c__U16String));
+    if (!data) {
+      return false;
+    }
+    // initialize all sequence elements
+    for (size_t i = 0; i < size; ++i) {
+      if (!rosidl_generator_c__U16String__init(&data[i])) {
+        /* free currently allocated and return false */
+        for (; i-- > 0; ) {
+          rosidl_generator_c__U16String__fini(&data[i]);
+        }
+        free(data);
+        return false;
+      }
+    }
+  }
+  sequence->data = data;
+  sequence->size = size;
+  sequence->capacity = size;
+  return true;
+}
+
+void
+rosidl_generator_c__U16String__Sequence__fini(
+  rosidl_generator_c__U16String__Sequence * sequence)
+{
+  if (!sequence) {
+    return;
+  }
+  if (sequence->data) {
+    // ensure that data and capacity values are consistent
+    assert(sequence->capacity > 0);
+    // finalize all sequence elements
+    for (size_t i = 0; i < sequence->capacity; ++i) {
+      rosidl_generator_c__U16String__fini(&sequence->data[i]);
+    }
+    free(sequence->data);
+    sequence->data = NULL;
+    sequence->size = 0;
+    sequence->capacity = 0;
+  } else {
+    // ensure that data, size, and capacity values are consistent
+    assert(0 == sequence->size);
+    assert(0 == sequence->capacity);
```

Comment:
@dirk-thomas here I see asserts, but for [rosidl_generator_c__U16String__fini](https://github.com/ros2/rosidl/pull/334/files#diff-be21fe7012d194da5384f06fbc5b7b72R39) I see `fprintf(...); exit(...);`. Any reason for both not receiving the same treatment?

Reply:
The code matches the one present for `String`. `U16String` replicates the exact same just for 16-bit characters.